teps/0011-redirecting-step-output-streams.md

```diff
@@ -0,0 +1,219 @@
+---
+title: redirecting-step-output-streams
+authors:
+  - "@chhsia0"
+creation-date: 2020-08-17
+last-updated: 2020-08-18
+status: proposed
+---
+
+# TEP-0011: Redirecting Step Output Streams
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Option 1: Allowing Users to Specify Redirection Paths](#option-1-allowing-users-to-specify-redirection-paths)
+    - [Example Usage](#example-usage)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Option 2: Redirecting to Canonical Conventional Paths](#option-2-redirecting-to-canonical-conventional-paths)
+    - [Example Usage](#example-usage-1)
+    - [Risks and Mitigations](#risks-and-mitigations-1)
+- [Design Details](#design-details)
+- [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Consuming outputs of a step in another step is a common pattern in writing Tasks. However, this is currently tedious to do. Task authors have to overwrite the image entrypoint with either `sh -c` or the `script` field to wrap up the command to run with an explicit output stream redirection. This is not even possible if the image does not come with a shell, e.g., images built by [`ko`](https://github.com/google/ko).
+
+To achieve the functionality of output stream redirection between steps and even Tasks, Tekton will add new fields to steps for Task authors to specify paths to redirect stdout/stderr to.
+
+## Motivation
+
+Allow Task authors to support the following use cases:
+
+* Run image `gcr.io/k8s-staging-boskos/boskosctl` with args in a step and process its output through another `jq` image in a subsequent step to acquire the project name without overwriting the image entrypoint, building custom image with both utilities, or looking into container logs ([tektoncd/pipeline#2925](https://github.com/tektoncd/pipeline/issues/2925)).
+
+* Run image `alpine/git` with args `["describe", "--tags"]` in a step and extract the stdout as a Task result without overwriting the image entrypoint to be used by subsequent tasks.
+
+* In general, enable Task authors to access stdout/stderr of a step easily in another step in the same or different Task, without requiring authors to: 1) overwrite image entrypoints; 2) write wrapping scripts solely to do output stream redirection; and 3) build wrapper images to include all needed utilities.
```

Comment:
None of these three points seem like terribly onerous requirements for Task authors. If they are, we can work to make them easier, rather than adding fields to the Tekton API surface.